File: src/java/sort/CountingSort.java
package sort;

/*
 * Sample implementation of Counting Sort
 *
 * https://www.geeksforgeeks.org/counting-sort/
 *
 * Counting sort makes assumptions about the data. It doesn't use comparisons and only works with positive discrete values. The values
 * must be within a specific range. It counts the number of occurrences of each value.
 *
 * Not in place algorithm
 * O(n) because of the assumptions of the data we make
 * This algorithm is for small range data
 */

public class CountingSort {
    public static void main(String[] args) {
        int[] arr = {2, 5, 9, 8, 2, 8, 7, 10, 4, 3};

        countingSort(arr, 1, 10);

        for (int i = 0; i <arr.length; i++) {
            System.out.println(arr[i]);
        }
    }

    public static void countingSort(int[] input, int min, int max) {
        int[] countArray = new int[(max-min) + 1];

        for (int i = 0; i < input.length; i++) {
            countArray[input[i] - min]++;
        }

        int j = 0;
        for (int i = min; i <= max; i++) {
            while (countArray[i-min] > 0) {
                input[j++] = i;
                countArray[i - min]--;
            }
        }
    }
}

File: src/c++/sort/CountingSort.cpp
#include <iostream>
#include <vector>
using namespace std;

void sort(vector<int>& arr, int k);

/*
 * COUNTING SORT ALGORITHM
 * 
 * All values are within the range of 1 to K and another array is used to count the frequencies of each value.
 * That frequency array is used to build up the new sorted array.
 *
 * O(N + K), stable, not in-place, non-comparison
 * 
 */

void sort(vector<int>& arr, int k) {
    vector<int> counts(k);
    for (int i = 0; i < arr.size(); i++) {
        counts[arr[i]-1]++;
    }

    int j = 0;
    for (int i = 0; i < arr.size(); i++) {
        if (counts[j] == 0) {
            j++;
            i--;
            continue;
        }
        arr[i] = j+1;
        counts[j]--;
    }
}

int main() {
    vector<int> arr = {3, 2, 5, 4, 4, 1, 3, 2, 5, 2, 4, 2}; // 1 2 2 2 2 3 3 4 4 4 5 5
    int k = 5;
    sort(arr, k);
    for (int i : arr) {
        cout << i << " ";
    }
    cout << endl;

    return 0;
}